infoblox/provider.go

```diff
@@ -21,36 +21,74 @@ const (
 	altIdSeparator      = "|"
 )
 
-func generateInternalId() string {
-	return uuid.NewString()
+type internalResourceId struct {
+	value uuid.UUID
 }
 
-func generateAltId(internalId string, ref string) string {
-	return fmt.Sprintf(
-		"%s%s%s",
-		internalId, altIdSeparator, ref)
+func (id *internalResourceId) Equal(id2 *internalResourceId) bool {
+	if id2 == nil {
+		panic("the argument must not be nil")
+	}
+	return id.value.String() == id2.value.String()
 }
 
-func getAltIdFields(altId string) (
-	internalId string, ref string, err error) {
+func (id *internalResourceId) String() string {
+	return id.value.String()
+}
 
-	idParts := strings.SplitN(altId, altIdSeparator, 2)
-	if len(idParts) != 2 {
-		err = fmt.Errorf("invalid internal ID for host record: '%s'", altId)
-		return
+// Returns a pointer to parsed internal resource ID, nil otherwise.
+func newInternalResourceIdFromString(id string) *internalResourceId {
+	newUUID, err := uuid.Parse(id)
+	if err != nil {
+		return nil
 	}
 
-	internalId = idParts[0]
-	ref = idParts[1]
-	return
+	return &internalResourceId{value: newUUID}
 }
 
-func renewAltId(oldAltId string, newRef string) (string, error) {
-	internalId, _, err := getAltIdFields(oldAltId)
+func generateInternalId() *internalResourceId {
+	uuid, err := uuid.NewRandom()
 	if err != nil {
-		return "", err
+		panic(err)
 	}
-	return generateAltId(internalId, newRef), nil
+	return &internalResourceId{value: uuid}
+}
+
+func isValidInternalId(internalId string) bool {
+	_, err := uuid.Parse(internalId)
+	if err != nil {
+		return false
+	}
+
+	return true
+}
+
+func generateAltId(internalId *internalResourceId, ref string) string {
+	if internalId == nil {
+		panic("the argument must not be nil")
+	}
+	return fmt.Sprintf(
+		"%s%s%s",
+		internalId.String(), altIdSeparator, ref)
+}
+
+// valid = true:
+//   - exactly 2 parts found
+//   - ... and separated by the delimiter
+//   - ... and the 1st one is a valid internal ID
+//   - ... and the 2nd one is not empty
+func getAltIdFields(altId string) (internalId *internalResourceId, ref string, valid bool) {
```

Comment:
Why don't you return the `error` value instead of `valid bool` value?

This will simplify the functions logic down the stack. Each time you're using the `valid` argument, you format the error in the caller function anyway ([here](https://github.com/skudriavtsev/terraform-provider-infoblox/blob/c93df1dbf76d7d498c2f882b4b348e1b5d28e0be/infoblox/resource_infoblox_ip_allocation.go#L262-L267), [here](https://github.com/skudriavtsev/terraform-provider-infoblox/blob/c93df1dbf76d7d498c2f882b4b348e1b5d28e0be/infoblox/resource_infoblox_ip_allocation.go#L447-L449), and [here](https://github.com/skudriavtsev/terraform-provider-infoblox/blob/c93df1dbf76d7d498c2f882b4b348e1b5d28e0be/infoblox/resource_infoblox_ip_allocation.go#L538-L540)).

Instead of formatting error in the caller function each time, just format it in `getAltIdFields`, and return instead of `valid` return argument.

Reply:
Probably you are right. I do not remember the exact reason. I will try to use the canonical way, and if there will be a real reason why to choose the opposite, I will document it with a comment.